File: 2221-check-if-a-parentheses-string-can-be-valid/check-if-a-parentheses-string-can-be-valid.js
/**
 * @param {string} s
 * @param {string} locked
 * @return {boolean}
 */
var canBeValid = function (s, locked) {
    const n = s.length;
    if (n % 2 != 0) {
        return false;
    }

    let count = 0;
    for (let i = 0; i < n; i++) {
        if (locked[i] === '0' || s[i] === '(') {
            count++;
        }
        else {
            count--;
        }
        if (count < 0) {
            return false;
        }
    }
    count = 0

    for (let i = n - 1; i >= 0; i--) {
        if (locked[i] === '0' || s[i] === ')') {
            count++;
        }
        else {
            count--;
        }
        if (count < 0) {
            return false;
        }
    }


    return true;
};

File: 2221-check-if-a-parentheses-string-can-be-valid/check-if-a-parentheses-string-can-be-valid.py
class Solution:
    def canBeValid(self, s: str, locked: str) -> bool:
        n = len(s)

        if n % 2 != 0:
            return False

        unlock = []
        openbr = []

        for i in range(n):
            if locked[i] == "0":
                unlock.append(i)
            elif s[i] == "(":
                openbr.append(i)
            elif s[i] == ")":
                if openbr:
                    openbr.pop()
                elif unlock:
                    unlock.pop()
                else:
                    return False

        while openbr and unlock and openbr[-1] < unlock[-1]:
            openbr.pop()
            unlock.pop()

        return False if openbr else True
